src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitNotificationEventRepository.kt

```diff
@@ -21,6 +22,26 @@ interface VisitNotificationEventRepository : JpaRepository<VisitNotificationEven
     notificationEvent: NotificationEventType,
   ): Boolean
 
+  /**
+   * The inner query will only return one True or no Result because of the HAVING COUNT(reference)=2 and LIMIT 1 the outer query will then
+   * return true if a result otherwise false. If we did not have the outer query we would get a null pointer.
+   *
+   * This is a very efficient way of doing these calculations using code will be inefficient
+   */
+  @Query(
+    "SELECT count(*) = 1 from (SELECT COUNT(*) = 2 FROM visit_notification_event vne " +
+      " WHERE vne.create_timestamp BETWEEN NOW() - INTERVAL '10 MINUTE' AND NOW() AND " +
+      "   vne.booking_reference in (:bookingReferencePair1,:bookingReferencePair2) AND " +
+      "   vne.type=:#{#notificationEvent.name()} " +
```

Comment:
what if the event is an non association update event and not a create event? Does the code need to differentiate between them?

Reply:
good point

Reply:
if (NON_ASSOCIATION_CREATED == notificationDto.type) {
      val prisonCode = prisonerService.getPrisonerSupportedPrisonCode(notificationDto.prisonerNumber)
      prisonCode?.let {
        val affectedVisits = getOverLappingVisits(notificationDto, prisonCode)
        processVisitsWithNotifications(affectedVisits, NON_ASSOCIATION_EVENT)
      }
    } else if (notificationDto.type in arrayOf(NON_ASSOCIATION_DELETED, NON_ASSOCIATION_CLOSED)) {